src/net_processing.cpp

```diff
@@ -3515,6 +3614,18 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
         pfrom.fSuccessfullyConnected = true;
+
+        if (pfrom.IsSensitiveRelayConn()) {
+            // The peer may intend to later send us NetMsgType::FEEFILTER limiting
+            // cheap transactions, but we don't wait for that and thus we may send
+            // them a transaction below their threshold. This is ok because this
+            // relay logic is designed to work even in cases when the peer drops
+            // the transaction (due to it being too cheap, or for other reasons).
+            PushUnbroadcastTx(pfrom, *peer, interruptMsgProc);
+
+            MaybeSendPing(pfrom, *peer, GetTime<std::chrono::microseconds>());
```

Comment:
This implementation of that logic doesn't seem ideal to me:

 * sending the TX without doing INV/GETDATA first risks having the TX be ignored as unrequested

 * `PushUnbroadcastTx` doesn't actually send the TX, but `MaybeSendPing` does actually send the PING, so you might get the PONG and closes the connection before the other end has actually received/processed the TX

Reply:
> sending the TX without doing INV/GETDATA first risks having the TX be ignored as unrequested

I thought about doing `INV` first and waiting for get data but decided to not do that for the following reasons:

1. Bitcoin Core accepts unsolicited transactions. According to https://developer.bitcoin.org/reference/p2p_networking.html#tx there is already some software that sends unsolicited transactions.
2. It would add complexity to the implementation.
3. After `INV` we would have to wait for some time for `GETDATA`, then in some cases when the wait is too little we risk cutting the connection just before we get `GETDATA`.
4. Last, but this is the most important IMO: it does not matter. There is redundancy: the transaction will be sent to more than one peer and we will retry if we don't hear back from the network about it. It is expected and is ok if some of the peers we send the transaction drops it.

> `PushUnbroadcastTx` doesn't actually send the TX, but `MaybeSendPing` does actually send the PING, so you might get the PONG and closes the connection before the other end has actually received/processed the TX

I don't think so. Both `PushUnbroadcastTx()` and `MaybeSendPing()` end up calling `PushMessage()` which enqueues the message in `CNode::vSendMsg`. This is done synchronously so the transaction will go before the ping. Then messages from `CNode::vSendMsg` are processed and sent in the order they were added (FIFO).